packages/toolkit/src/query/core/buildSlice.ts

```diff
@@ -156,11 +158,37 @@ export function buildSlice({
             meta.arg.queryCacheKey,
             (substate) => {
               if (substate.requestId !== meta.requestId) return
+              const { merge } = definitions[
+                meta.arg.endpointName
+              ] as QueryDefinition<any, any, any, any>
               substate.status = QueryStatus.fulfilled
-              substate.data =
-                definitions[meta.arg.endpointName].structuralSharing ?? true
-                  ? copyWithStructuralSharing(substate.data, payload)
-                  : payload
+
+              if (merge) {
+                if (substate.data !== undefined) {
+                  // There's existing cache data. Let the user merge it in themselves.
+                  // We're already inside an Immer-powered reducer, and the user could just mutate `substate.data`
+                  // themselves inside of `merge()`. But, they might also want to return a new value.
+                  // Try to let Immer figure that part out, save the result, and assign it to `substate.data`.
+                  let newData = createNextState(
```

Comment:
The nested `createNextState` is not causing any problems? I do remember *something* vaguely that this had caused problems in the past.

Reply:
Yeah, Immer had some issues with nested drafts in earlier versions. That behavior has been made more consistent now.  I did some testing here with both mutate and return to make sure they work, and also confirmed it throws if you mutate + return.